src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,483 @@
+         */
+        public Builder addIKM(byte[] ikm) {
+            Objects.requireNonNull(ikm, "ikm must not be null");
+            if (ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "Generic"));
+            } else {
+                return this;
+            }
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * An implementation should concatenate the salt into a single value
+         * once all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(SecretKey salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            salts.add(salt);
+            return this;
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * An implementation should concatenate the salt into a single value
+         * once all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
+     * {@code ExtractThenExpand} objects.
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Creates an {@code Expand} object
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null}
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte array is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output key material (must be greater than 0 and
+     *     less than 255 * HMAC length)
```

Comment:
The maximum size of `length` is not checked in this class but it's worth mentioning. Also, the size of `prk` also has a minimum size that is not checked here. I suggest talking about both in the method spec.

Reply:
The way this is worded it is part of the API specification, so I would expect `IllegalArgumentException` to be thrown if the length is less than 255 * HMAC length.  But as I understand, only the HKDF impl can do that. I would change this to only say the length must be greater than 0, and add an "Implementation Note" that says HKDF implementations will enforce that the length is less than 255 * HMAC length.

Reply:
Thanks for the update. The `prk` one also needs to be mentioned. I should have added a comment yesterday.